editor/scene_tree_dock.cpp

```diff
@@ -2493,6 +2507,79 @@ void SceneTreeDock::replace_node(Node *p_node, Node *p_by_node, bool p_keep_prop
 	}
 }
 
+void SceneTreeDock::perform_node_replace(Node *p_base, Node *p_node, Node *p_by_node) {
+	if (!p_base) {
+		p_base = edited_scene;
+	}
+
+	if (!p_base) {
+		return;
+	}
+
+	// Renaming node used in node properties.
+	List<PropertyInfo> properties;
+	p_base->get_property_list(&properties);
+
+	for (const PropertyInfo &E : properties) {
+		if (!(E.usage & (PROPERTY_USAGE_STORAGE | PROPERTY_USAGE_EDITOR))) {
+			continue;
+		}
+		String propertyname = E.name;
+		Variant old_variant = p_base->get(propertyname);
+		Variant updated_variant = old_variant;
+
+		if (_check_node_recursive(updated_variant, p_node, p_by_node, E.hint_string)) {
+			EditorUndoRedoManager *undo_redo = EditorUndoRedoManager::get_singleton();
+			undo_redo->add_do_property(p_base, propertyname, updated_variant);
+			undo_redo->add_undo_property(p_base, propertyname, old_variant);
+			p_base->set(propertyname, updated_variant);
+		}
+	}
+
+	for (int i = 0; i < p_base->get_child_count(); i++) {
+		perform_node_replace(p_base->get_child(i), p_node, p_by_node);
+	}
+}
+
+bool SceneTreeDock::_check_node_recursive(Variant &r_variant, Node *p_node, Node *p_by_node, const String type_hint) {
+	switch (r_variant.get_type()) {
+		case Variant::OBJECT: {
+			if (p_node == r_variant) {
+				if (p_by_node->is_class(type_hint) || EditorNode::get_singleton()->is_object_of_custom_type(p_by_node, type_hint)) {
+					r_variant = p_by_node;
+					return true;
+				} else {
+					r_variant = memnew(Object);
+					ERR_FAIL_V_MSG(true, vformat(TTR("The node which type was changed was stored in a variable. This variable needed a node of type %s but the new type is %s. Removing the node from this variable."), type_hint, p_by_node->get_class()));
```

Comment:
I'm not confident about my wording at all so pleas propose better wording for explaining what happens

Reply:
maybe this could be a warning instead of an error?

Reply:
```suggestion
					ERR_FAIL_V_MSG(true, vformat(TTR("The node's new type is incompatible with an exported variable (expected %s, but type is %s). Removing the node from this variable."), type_hint, p_by_node->get_class()));
```
I agree it should be a warning. Also you could mention the variable name if possible.